common/route/awooter/awooter.cc

```diff
@@ -183,63 +171,46 @@ extern "C" {
         return dst_loc;
     }
 
-    uint64_t npnr_context_get_pips_leak(const Context *const ctx, PipId **pips) {
-        auto pip_vec = std::vector<PipId>{};
+    size_t npnr_context_get_pips_leak(const Context *const ctx, rust::Vec<WireId> *pips) {
         for (auto pip : ctx->getPips()) {
-            pip_vec.push_back(pip);
+            pips.push_back(pip);
         }
-        pip_vec.shrink_to_fit();
-        auto size = pip_vec.size();
-        *pips = pip_vec.data();
-        auto dummy = std::vector<PipId>{};
-        // Yes, by memcpying over `pip_vec` we leak memory.
-        std::memcpy(&pip_vec, &dummy, sizeof(dummy));
-        return size;
+        pips.shrink_to_fit();
+        return pips.size();
     }
 
-    uint64_t npnr_context_get_wires_leak(const Context *const ctx, WireId **wires) {
-        auto wire_vec = std::vector<WireId>{};
+    size_t npnr_context_get_wires_leak(const Context *const ctx, rust::Vec<WireId> *wires) {
         for (auto wire : ctx->getWires()) {
-            wire_vec.push_back(wire);
+            wires.push_back(wire);
         }
-        wire_vec.shrink_to_fit();
-        auto size = wire_vec.size();
-        *wires = wire_vec.data();
-        auto dummy = std::vector<WireId>{};
-        // Yes, by memcpying over `wire_vec` we leak memory.
-        std::memcpy(&wire_vec, &dummy, sizeof(dummy));
-        return size;
+        wires.shrink_to_fit();
+        return wires.size();
     }
 
+    size_t npnr_netinfo_users_leak(NetInfo &net, rust::Vec<PortRef> &vec) {
+        for (auto& item : net.users) {
+            vec.push_back(item);
+        }
+        vec.shrink_to_fit();
+        return vec.size();
+    }
```

Comment:
these no longer leak memory, i would recommend renaming them here and in rust accordingly